### PR TITLE
CDAP-4592 Schema indexing with full schema body

### PR DESCRIPTION
**Pull request by rohitsinha54**
- Introduce Indexer interface for indexing MetadataEntry
- SchemaIndexer to index schema value rather than split entries as values
- DefaultValueIndexer for indexing values by default
- IndexerFactory to create required Indexer

Issue: https://issues.cask.co/browse/CDAP-4592
Build: http://builds.cask.co/browse/CDAP-RBT603-3

@bdmogal @poornachandra : Please review. Thanks.

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 1]
```diff
@@ -96,7 +98,20 @@ private void setMetadata(MetadataEntry metadataEntry) {
...
    */
   public void setProperty(Id.NamespacedId targetId, String key, String value) {
-    setMetadata(new MetadataEntry(targetId, key, value));
+    setProperty(targetId, key, value, IndexerFactory.getIndexer(IndexerType.DEFAULT_VALUE_INDEXER));
+  }
+
+  /**
+   * Sets a metadata property for the specified {@link Id.NamespacedId}.
+   *
+   * @param targetId The target Id: app-id(ns+app) / program-id(ns+app+pgtype+pgm) /
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: I don't think you need the info in parentheses.
Simply specifying `ProgramId` means ns + app + pgtype + pgmName, because of how `ProgramId` class is defined, right?

Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: I don't think a little extra details hurts. Specially, if I don't have type it ;)

Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Extra detail will hurt when implementation of `ProgramId` changes. Instead use `@link` to link to `ProgramId` class.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/DefaultValueIndexer.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * Default {@link Indexer} for {@link MetadataEntry}
+ */
+public class DefaultValueIndexer implements Indexer {
+  private static final Pattern VALUE_SPLIT_PATTERN = Pattern.compile("[-_:\\s]+");
+  private static final Pattern TAGS_SEPARATOR_PATTERN = Pattern.compile("[,\\s]+");
+
+  @Override
+  public Set<String> getIndexes(MetadataEntry entry) {
+    Set<String> valueIndexes = new HashSet<>();
+    if (entry.getValue().contains(MetadataDataset.TAGS_SEPARATOR)) {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/DefaultValueIndexer.java]: Instead of checking for presence of `TAGS_SEPARATOR` to determine if metadata is tag, why don't we just check if key equals `tags`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      return Schema.parseJson(schemaStr);
+    } catch (IOException e) {
+      throw new IllegalArgumentException(e);
+    }
+  }
+
+  private Set<String> createIndexes(Schema schema) {
+    Set<String> indexes = new HashSet<>();
+    if (schema == null) {
+      return indexes;
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: If schema is null, just return `Collections.empytSet()`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    if (schema.isSimpleOrNullableSimple()) {
+      Schema.Type type = getSimpleType(schema);
+      // put the type as index
+      indexes.add(type.toString());
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: In this case just return `Collections.singleton()`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    } else {
+      for (Schema.Field field : schema.getFields()) {
+        String fieldName = field.getName();
+        // TODO: What if field.getSchema() is not simple or nullable simple?
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: If schema is not simple type then we have to add the components of that schema, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java [hunk 6]
```diff
@@ -60,6 +62,14 @@ public AbstractSystemMetadataWriter(MetadataStore metadataStore, Id.NamespacedId
   abstract String[] getSystemTagsToAdd();
 
   /**
+   * Define the {@link MetadataScope#SYSTEM system} schema to add for this entity.
+   *
+   * @return the schema as a {@link String}
+   */
+  @Nullable
+  abstract String getSchemaToAdd();
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: Since this method can return null, give a default implementation that returns null. This will prevent derived classes that don't return schema from implementing this method.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/IndexerFactory.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package co.cask.cdap.data2.metadata.indexer;
+
+import com.google.common.base.Throwables;
+
+/**
+ * A factory for {@link Indexer}
+ */
+public class IndexerFactory {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/IndexerFactory.java]: Since there is no special logic necessary to instantiate an indexer, there is no need for factory, right? We can just new the classes to instantiate them. I don't see the reason for a factory, nor the reason for `IndexerType` enum. Let's keep implementation simple as far as possible.

**Comment by poornachandra** [comment]
@rohitsinha54 I have taken a pass over it. Please address comments.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        String fieldType = getSimpleType(field.getSchema()).toString();
+        indexes.add(fieldName + MetadataDataset.KEYVALUE_SEPARATOR + fieldType);
+        indexes.add(fieldName);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: For completeness' sake let's index just field type too.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java (hunk 3 above)
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: Nice

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java (hunk 5 above)
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: I think the idea was to support simple types. This is moved from the system metadata pr. Lets ask @bdmogal .

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/IndexerFactory.java (hunk 7 above)
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/IndexerFactory.java]: I will change it to just new but my thoughts were to have a factory which can provide indexers to everyone rather than newing different indexers all over the place.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    // otherwise its a json
+    try {
+      return Schema.parseJson(schemaStr);
+    } catch (IOException e) {
+      throw new IllegalArgumentException(e);
+    }
+  }
+
+  private Set<String> createIndexes(Schema schema) {
+    Set<String> indexes = new HashSet<>();
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: This initialization can be moved to the else block, right? Then you don't create it when schema is empty or simple.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java (hunk 5 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: The reason I'm saying let's index everything is because of the following -
If there is a schema like -
```
{ 
   type : int,
   emp: record
      {
          name: string,
          id: string
      }
}
```
name and id are simple types, right? Do we index them?

**Comment by poornachandra** [comment]
@rohitsinha54 Note that I've replied to your hidden comment https://github.com/caskdata/cdap/pull/4924#discussion_r49951679

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java (hunk 5 above)
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: @poornachandra : This is possible only in cases of Records. How we will index others such Map, Lists etc. If I remember correctly we decided to support simple types for now but if we want to support complex then can you suggest for others too or you are suggesting just to do it for Record only ?

Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: We discussed this offline. We decided to index both lists and maps too.

Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: So I started working on this to implement it the way we discussed. For complex type the type and then its internal type.
I looked into this complex schema https://github.com/caskdata/cdap/blob/5159926306b0c9db8483adf73fa9fa6fc9a40640/cdap-common/src/test/java/co/cask/cdap/io/SchemaTest.java#L182-L182
It looks like support for complex and nested schema will need some work to cover all the edge cases. So I propose to merge this PR tonight and then we can revisit this to support nested complex schema tomorrow.

Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/indexer/SchemaIndexer.java]: Sounds good.

**Comment by poornachandra** [comment]
LGTM

**Comment by rohitsinha54** [comment]
Build passed here: http://builds.cask.co/browse/CDAP-RBT603-4 Merging.
